libdnf5-plugins/expired-pgp-keys/expired-pgp-keys.cpp

```diff
@@ -0,0 +1,236 @@
+        for (size_t i = 0; attrs[i]; ++i) {
+            if (std::strcmp(attribute, attrs[i]) == 0) {
+                return attrs_value[i];
+            }
+        }
+        return nullptr;
+    }
+
+    void resolved(const libdnf5::base::Transaction & transaction) override { process_expired_pgp_keys(transaction); }
+
+private:
+    void process_expired_pgp_keys(const libdnf5::base::Transaction & transaction) const;
+};
+
+/// Check that GPG is installed to enable querying expired keys later.
+static bool is_gpg_installed() {
+    auto ts = rpmtsCreate();
+    rpmdbMatchIterator mi;
+    mi = rpmtsInitIterator(ts, RPMDBI_PROVIDENAME, "gpg", 0);
+    bool found = rpmdbNextIterator(mi) != NULL;
+    rpmdbFreeIterator(mi);
+    rpmtsFree(ts);
+    return found;
+}
+
+/// Check if the transaction contains any inbound actions.
+/// This determines if new software is to be installed, which might require downloading a new PGP signing key.
+static bool any_inbound_action(const libdnf5::base::Transaction & transaction) {
+    for (const auto & package : transaction.get_transaction_packages()) {
+        if (transaction_item_action_is_inbound(package.get_action())) {
+            return true;
+        }
+    }
+    return false;
+}
+
+/// Retrieve the PGP key expiration timestamp, or return -1 if the expiration is not available.
+static int64_t get_key_expire_timestamp(std::string raw_key) {
+    // open gpg process to retrieve information about the key
+    std::unique_ptr<FILE, int (*)(FILE *)> pipe(
+        popen(libdnf5::utils::sformat("echo \"{}\" | gpg --show-keys --with-colon", raw_key).c_str(), "r"), pclose);
+    if (!pipe) {
+        return -1;
+    }
+
+    // read key information from the gpg process
+    char buffer[1024];
+    std::string output;
+    while (fgets(buffer, sizeof(buffer), pipe.get()) != nullptr) {
+        output += buffer;
+    }
+
+    // check expired time is a numeric value
+    auto expired_date_string = libdnf5::utils::string::split(libdnf5::utils::string::split(output, "\n")[0], ":")[6];
+    if (expired_date_string.empty() || expired_date_string.find_first_not_of("0123456789") != std::string::npos) {
+        return -1;
+    }
+
+    return static_cast<int64_t>(std::stoull(expired_date_string));
+}
+
+/// Remove the system package corresponding to the PGP key from the given RPM header.
+static bool remove_pgp_key(const libdnf5::rpm::KeyInfo & key) {
+    bool retval{false};
+    auto ts = rpmtsCreate();
+    Header h;
+    rpmdbMatchIterator mi;
+    mi = rpmtsInitIterator(ts, RPMDBI_NAME, "gpg-pubkey", 0);
+    auto key_id = key.get_short_key_id();
+    while ((h = rpmdbNextIterator(mi)) != nullptr) {
+        if (headerGetAsString(h, RPMTAG_VERSION) == libdnf5::utils::string::tolower(key_id)) {
+            rpmtsAddEraseElement(ts, h, -1);
+            retval = rpmtsRun(ts, nullptr, RPMPROB_FILTER_NONE) == 0;
+            break;
+        }
+    }
+    rpmtsFree(ts);
+    rpmdbFreeIterator(mi);
```

Comment:
I'm not sure if this is mandatory, but at least in general it's a good idea to release the iterators first and then the transaction.